crates/intern/src/gc.rs

```diff
@@ -0,0 +1,325 @@
+                gc.retain_only_alive(&mut *shard.write(), |item| item.0.as_ptr().addr())
+            });
+        } else {
+            storage.shards().par_iter().for_each(|shard| {
+                gc.retain_only_alive(&mut *shard.write(), |item| item.0.as_ptr().addr())
+            });
+        }
+    }
+}
+
+pub trait GcInternedVisit {
+    fn visit_with(&self, gc: &mut GarbageCollector);
+}
+
+pub trait GcInternedSliceVisit: SliceInternable {
+    fn visit_header(header: &Self::Header, gc: &mut GarbageCollector);
+    fn visit_slice(header: &[Self::SliceType], gc: &mut GarbageCollector);
+}
+
+#[derive(Default)]
+pub struct GarbageCollector {
+    alive: FxHashSet<usize>,
+    storages: Vec<Box<dyn Storage + Send + Sync>>,
+}
+
+impl GarbageCollector {
+    pub fn add_storage<T: Internable + GcInternedVisit>(&mut self) {
+        const { assert!(T::USE_GC) };
+
+        self.storages.push(Box::new(InternedStorage::<T>(PhantomData)));
+    }
+
+    pub fn add_slice_storage<T: SliceInternable + GcInternedSliceVisit>(&mut self) {
+        const { assert!(T::USE_GC) };
+
+        self.storages.push(Box::new(InternedSliceStorage::<T>(PhantomData)));
+    }
+
+    /// # Safety
+    ///
+    /// This cannot be called if there are some not-yet-recorded type values.
+    pub unsafe fn collect(mut self) {
+        let total_nodes = self.storages.iter().map(|storage| storage.len()).sum();
+        self.alive = FxHashSet::with_capacity_and_hasher(total_nodes, FxBuildHasher);
+
+        let storages = std::mem::take(&mut self.storages);
+
+        for storage in &storages {
+            storage.mark(&mut self);
+        }
+
+        if cfg!(miri) {
+            storages.iter().for_each(|storage| storage.sweep(&self));
+        } else {
+            storages.par_iter().for_each(|storage| storage.sweep(&self));
+        }
+    }
+
+    pub fn mark_interned_alive<T: Internable>(
+        &mut self,
+        interned: InternedRef<'_, T>,
+    ) -> ControlFlow<()> {
+        if interned.strong_count() > 1 {
+            // It will be visited anyway, so short-circuit
+            return ControlFlow::Break(());
+        }
+        let addr = interned.as_raw().addr();
+        if !self.alive.insert(addr) { ControlFlow::Break(()) } else { ControlFlow::Continue(()) }
+    }
+
+    pub fn mark_interned_slice_alive<T: SliceInternable>(
+        &mut self,
+        interned: InternedSliceRef<'_, T>,
+    ) -> ControlFlow<()> {
+        if interned.strong_count() > 1 {
+            // It will be visited anyway, so short-circuit
+            return ControlFlow::Break(());
+        }
+        let addr = interned.as_raw().addr();
+        if !self.alive.insert(addr) { ControlFlow::Break(()) } else { ControlFlow::Continue(()) }
+    }
+
+    #[inline]
+    fn retain_only_alive<T>(&self, map: &mut RawTable<T>, mut get_addr: impl FnMut(&T) -> usize) {
+        unsafe {
+            // Here we only use `iter` as a temporary, preventing use-after-free
```

Comment:
I don't understand this comment

Reply:
The code was copied from `DashMap`'s `retain()` (which I can't use because does not run in parallel), the comment is from there too.